### PR TITLE
feat: Abstract LLM/VLM forward-backward step

### PR DESCRIPTION
**Pull request by HuiyingLi**
## Summary

- Add a shared forward/backward helper for LLM and VLM fine-tuning paths.
- Route LLM and VLM recipes through the shared helper while keeping recipe-specific hooks for CP prep, FP8 context, and VLM PP media staging.
- Enable VLM pipeline-parallel validation by preparing validation dataloader media chunks for PP and running validation through `schedule.eval`.

## Why

The LLM and VLM recipes had largely duplicated forward/backward control flow. VLM PP validation was previously skipped because validation used a direct model-forward path and did not prepare VLM media tensors for PP microbatches.

## Validation

- `python -m py_compile nemo_automodel/components/training/forward_backward.py nemo_automodel/recipes/llm/train_ft.py nemo_automodel/recipes/vlm/finetune.py`
- `python -m ruff check nemo_automodel/components/training/forward_backward.py nemo_automodel/recipes/llm/train_ft.py nemo_automodel/recipes/vlm/finetune.py`
- `git diff --check`
- `pytest -q tests/unit_tests/recipes/test_train_ft.py`
- `pytest -q tests/unit_tests/recipes/test_finetune_vlm_helpers.py`
- Local Qwen3 VL MoE 30B EP4/PP2 smoke run with validation:
  - train: `loss 2.3266`, `num_label_tokens 930`
  - val: `loss 2.2712`, `num_label_tokens 941`

### CODE REVIEW

File: nemo_automodel/components/training/forward_backward.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    *,
+    idx: int,
+    batch: MutableMapping[str, Any],
+    device: torch.device,
+    device_mesh: Any,
+    model_parts: list[torch.nn.Module],
+    distributed_config: Any,
+    loss_fn: Callable[..., torch.Tensor] | None,
+    calculate_loss_fn: Callable[..., torch.Tensor],
+    loss_buffer: list[torch.Tensor],
```
Comment (akoumpa) [inline nemo_automodel/components/training/forward_backward.py]: I would let the called handle this, and instead return the loss to the caller

File: nemo_automodel/components/training/forward_backward.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+def forward_backward_step(
+    *,
+    idx: int,
+    batch: MutableMapping[str, Any],
+    device: torch.device,
+    device_mesh: Any,
+    model_parts: list[torch.nn.Module],
+    distributed_config: Any,
+    loss_fn: Callable[..., torch.Tensor] | None,
+    calculate_loss_fn: Callable[..., torch.Tensor],
```
Comment (akoumpa) [inline nemo_automodel/components/training/forward_backward.py]: can we consolidate the two
- loss_fn
- calculate_loss_fn
?

File: nemo_automodel/components/training/forward_backward.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    loss_buffer: list[torch.Tensor],
+    num_label_tokens: int | None,
+    num_batches: int,
+    is_train: bool,
+    pp_enabled: bool,
+    pp: Any | None,
```
Comment (akoumpa) [inline nemo_automodel/components/training/forward_backward.py]: is there a strong use-case for pp_enabled? if not, i would do something like `pp_enabled = (pp is not None)` inside `forward_backward_step`'s body.

File: nemo_automodel/components/training/forward_backward.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    dp_group_size: int,
```
Comment (akoumpa) [inline nemo_automodel/components/training/forward_backward.py]: can we derive this from device_mesh by placing a convention on mesh naming?

File: nemo_automodel/components/training/forward_backward.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    if isinstance(value, dict):
+        return {key: move_to_device(item, device) if item is not None else None for key, item in value.items()}
+    if isinstance(value, list):
+        return [move_to_device(item, device) for item in value]
+    if isinstance(value, tuple):
+        return tuple(move_to_device(item, device) for item in value)
+    return value
+
+
+def forward_backward_step(
```
Comment (akoumpa) [inline nemo_automodel/components/training/forward_backward.py:99]: I feel this function is challenging to get to the right abstraction, because it is trying to abstract LLM/VLM + train/eval + PP/non-PP + CP + FP8 + FSDP sync + loss calculation + modality hooks, so I'm thinking what ways could be explored to simplify it, without losing critical functionality.
